src/visitors/semantic/completion/mod.rs

```diff
@@ -0,0 +1,152 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::protocol::properties::Position;
+use crate::protocol::responses::{
+    CompletionItem, CompletionItemKind, InsertTextFormat,
+};
+use crate::stdlib::Completable;
+use crate::visitors::semantic::utils;
+
+use flux::semantic::nodes::*;
+use flux::semantic::types::MonoType;
+use flux::semantic::walk::{Node, Visitor};
+
+#[derive(Default)]
+pub struct CompletableFinderState {
+    pub completables: Vec<Rc<dyn Completable>>,
+}
+
+pub struct CompletableFinderVisitor {
+    pub pos: Position,
+    pub state: Rc<RefCell<CompletableFinderState>>,
+}
+
+impl CompletableFinderVisitor {
+    pub fn new(pos: Position) -> Self {
+        CompletableFinderVisitor {
+            state: Rc::new(RefCell::new(
+                CompletableFinderState::default(),
+            )),
+            pos,
+        }
+    }
+}
+
+fn get_var_type(expr: Expression) -> Option<VarType> {
+    match expr {
+        Expression::Integer(_) => Some(VarType::Int),
+        Expression::Boolean(_) => Some(VarType::Bool),
+        Expression::Float(_) => Some(VarType::Float),
+        Expression::StringLit(_) => Some(VarType::String),
+        Expression::Array(_) => Some(VarType::Array),
+        Expression::Regexp(_) => Some(VarType::Regexp),
+        Expression::Duration(_) => Some(VarType::Duration),
+        Expression::Call(c) => {
+            let result_type = utils::follow_function_pipes(c);
+
+            match result_type {
+                MonoType::Int => Some(VarType::Int),
+                MonoType::Float => Some(VarType::Float),
+                MonoType::Bool => Some(VarType::Bool),
+                MonoType::Arr(_) => Some(VarType::Array),
+                MonoType::Duration => Some(VarType::Duration),
+                MonoType::Row(_) => Some(VarType::Row),
+                MonoType::String => Some(VarType::String),
+                MonoType::Uint => Some(VarType::Uint),
+                MonoType::Time => Some(VarType::Time),
+                _ => None,
+            }
+        }
+        _ => None,
+    }
+}
+
+impl<'a> Visitor<'a> for CompletableFinderVisitor {
+    fn visit(&mut self, node: Rc<Node<'a>>) -> bool {
+        let mut state = self.state.borrow_mut();
+        let loc = node.loc();
+        let pos = self.pos.clone();
```

Comment:
you can borrow here instead of copy

Reply:
Ownership with these visitors using walk in libflux can get tricky, prefer to clone than untangle ownership